### PR TITLE
Added Z-function and fuzz test

### PR DESCRIPTION
**Pull request by Chillee**
Featuring a fuzz-test now (thanks to Ludo Pulles).

### CODE REVIEW

File: content/strings/Zfunc.h [hunk 1]
```diff
@@ -0,0 +1,22 @@
+
+/**
+ * Author: chilli
+ * License: CC0
+ * Description: z[x] computes the length of the longest common prefix of s[i:] and s. (abacaba -> 0010301)
+ * Time: O(n)
+ * Status: fuzz-tested (thanks to Ludo Pulles)
```
Comment (simonlindholm) [inline content/strings/Zfunc.h]: move this thanks-note to the fuzz-test

File: content/strings/Zfunc.h [hunk 2]
```diff
@@ -0,0 +1,22 @@
+
```
Comment (simonlindholm) [inline content/strings/Zfunc.h]: remove blank line

File: content/strings/Zfunc.h [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+/**
+ * Author: chilli
+ * License: CC0
+ * Description: z[x] computes the length of the longest common prefix of s[i:] and s. (abacaba -> 0010301)
```
Comment (simonlindholm) [inline content/strings/Zfunc.h]: ", except z[0] = 0"?

File: content/strings/Zfunc.h [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+ * Time: O(n)
+ * Status: fuzz-tested (thanks to Ludo Pulles)
+ */
+#pragma once
+
+vi Z(string S) {
+	vi z(sz(S));
+	int l = -1, r = -1;
+	rep(i,1,sz(S)){
```
Comment (simonlindholm) [inline content/strings/Zfunc.h]: space between ) and {

File: content/strings/Zfunc.h [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+		z[i] = i >= r ? 0 : min(r - i, z[i - l]);
+		while (i + z[i] < sz(S) && S[i + z[i]] == S[z[i]])
+			z[i]++;
+		if (i + z[i] > r)
+			l = i, r = i + z[i];
+	}
+	return z;
+}
```
Comment (simonlindholm) [inline content/strings/Zfunc.h]: newline at end of file

**Comment by Chillee** [comment]
Responded to comments.

**Comment by Chillee** [comment]
Should I put the conditionals/loops on the same line?

**Comment by simonlindholm** [comment]
> Should I put the conditionals/loops on the same line?

I'm fine with it either way.

**Pull request merged**